gem/unconcatenate.py

```diff
@@ -0,0 +1,192 @@
+    shapes = [(other_index.extent,) for other_index in other_indices]
+
+    offset = 0
+    for multiindex in multiindices:
+        shape = tuple(index.extent for index in multiindex)
+        size = numpy.prod(shape, dtype=int)
+        slice_ = slice(offset, offset + size)
+        offset += size
+
+        sub_ref = Indexed(reshape(view(data, slice_, *slices),
+                                  shape, *shapes),
+                          multiindex + other_indices)
+        sub_ref, = remove_componenttensors((sub_ref,))
+        yield sub_ref
+
+
+def _replace_node(node, self):
+    """Replace subexpressions using a given mapping.
+
+    :param node: root of expression
+    :param self: function for recursive calls
+    """
+    assert isinstance(node, Node)
+    if self.cut(node):
+        return node
+    try:
+        return self.mapping[node]
+    except KeyError:
+        return reuse_if_untouched(node, self)
+
+
+def replace_node(expression, mapping, cut=None):
+    """Replace subexpressions using a given mapping.
+
+    :param expression: a GEM expression
+    :param mapping: a :py:class:`dict` containing the substitutions
+    :param cut: cutting predicate; if returns true, it is assumed that
+                no replacements would take place in the subexpression.
+    """
+    mapper = Memoizer(_replace_node)
+    mapper.mapping = mapping
+    mapper.cut = cut or (lambda node: False)
+    return mapper(expression)
+
+
+def _unconcatenate(cache, pairs):
+    # Tail-call recursive core of unconcatenate.
+    # Assumes that input has already been sanitised.
+    concat_group = find_group([e for v, e in pairs])
+    if concat_group is None:
+        return pairs
+
+    # Get the index split
+    concat_ref = next(iter(concat_group))
+    assert isinstance(concat_ref, Indexed)
+    concat_expr, = concat_ref.children
+    index, = concat_ref.multiindex
+    assert isinstance(concat_expr, Concatenate)
+    try:
+        multiindices = cache[index]
+    except KeyError:
+        multiindices = tuple(tuple(Index(extent=d) for d in child.shape)
+                             for child in concat_expr.children)
+        cache[index] = multiindices
+
+    def cut(node):
+        """No need to rebuild expression of independent of the
+        relevant concatenation index."""
+        return index not in node.free_indices
+
+    # Build Concatenate node replacement mappings
+    mappings = [{} for i in range(len(multiindices))]
+    for concat_ref in concat_group:
+        concat_expr, = concat_ref.children
+        for i in range(len(multiindices)):
+            sub_ref = Indexed(concat_expr.children[i], multiindices[i])
+            sub_ref, = remove_componenttensors((sub_ref,))
+            mappings[i][concat_ref] = sub_ref
+
+    # Finally, split assignment pairs
+    split_pairs = []
+    for var, expr in pairs:
+        if index not in var.free_indices:
+            split_pairs.append((var, expr))
```

Comment:
It does happen during coefficient evaluation, but that also contains within the `IndexSum` a corresponding kernel argument that needs to be split. So it is always pairs of flexibly indexed variable and expression containing the `Concatenate` node.